### ISSUE TITLE
Stripping control codes

### ISSUE THREAD

**Issue by spreadred** [sopel-irc/sopel#462]
I am trying to write a bot using `@rule`s to trigger certain functions based on channel text. mIRC control codes seem to interfere with this. Is there a way an option to get willie to strip all mIRC control codes before passing the text to `@rule`?

**Comment by Exirel** [comment]
Are mIRC control codes still a thing today?

**Comment by kaptainkommie** [comment]
@Exirel mIRC is still alive and well, although not under active development IIRC. If you happen to find yourself in channels dedicated to serving files via XDCC and/or other methods, you likely will encounter these control codes, which are used to make their "pack listings" stand out. Although many modern clients seem to transparently strip them.

**Comment by Exirel** [comment]
Yeah, I saw the latest release of mIRC is from 2018, so it's alive, that's nice! I see that @dgw used some regex in #34, so maybe that could be reused to implement a `@rule(..., skip_control_code=True)`?

**Comment by dgw** [comment]
@Exirel That should be sopel-irc/sopel-extras#34, but yes. :grin:

I'm actually going to retitle the issue, because referring to them as "mIRC control codes" is no longer accurate. While mIRC may have begun the trend toward formatting support for IRC, it's far from the only client to recognize formatting characters today. #1302 tracked implementation of all remaining (not yet supported by `sopel.tools.formatting` methods) codes known to exist in at least one client, as documented [by modern.ircdocs.horse](https://modern.ircdocs.horse/formatting.html) at the time.

**Comment by Exirel** [comment]
Hm, in this case, I even wonder if that `skip_control_code` argument could be `False` in 7.x, then `True` in 8.x if the usage is so common? I mean, that looks reasonable to me.

**Comment by dgw** [comment]
Implementation thought: https://github.com/sopel-irc/sopel/issues/1768#issuecomment-625857898

**Comment by dgw** [comment]
After implementation discussion on IRC, we've (@Exirel and I) decided that it'll be cleaner to do this in 8.0 after we no longer have to worry about Python 2 support.

**Comment by dgw** [comment]
We are now in the 8.0 dev cycle and firmly located in py3-only land. This can be implemented with the use of keyword-only arguments to the relevant decorators now.

**Comment by Exirel** [comment]
Oh no. I fall into the rabbit hole:

* `plugin.rule` is one thing... but they are others: `find` and `search` for instance
* lazy versions are painful to deal with, since the current loader spec is to just return a list of patterns, no mention of control code or anything like that
* as pointed out by @dgw someone will, at some point, want to match both with and without control code for the same pattern, so making it really flexible is not easy at all

However, I've a counter offer to make:

```python
from sopel import plugin

def _rule_handler(bot, trigger):
    ...

@plugin.plain  # or @plugin.plain(True)
@plugin.rule('.*')
def plain_trigger(bot, trigger):
    _rule_handler(bot, trigger)

@plugin.plain(False)  # the default
@plugin.rule('.*')
def plain_trigger(bot, trigger):
    _rule_handler(bot, trigger)
```

If someone is **really** into having both version at the same time, they always can. Making `plain` a single flag on a plugin callable makes everything more simple and "just work" with very few changes.

Sure, that completely throw away the "but we have to wait until Python 3+ only" thing and also it forces plugin authors to create wrappers if they want both. On the other hand... what are the odd, exactly, to match both version?

... but wait! I've another idea: instead of a boolean flag, `plain` (or whatever the flag will be), could be more than a binary value. For instance, what if:

```python

@plugin.plain(plugin.PLAIN_ONLY)
def plain_trigger(bot, trigger):
    ...

@plugin.plain(plugin.RAW_ONLY)  # the default for now
def raw_trigger(bot, trigger):
    ...

@plugin.plain(plugin.PLAIN_OR_RAW)
def plain_or_raw_trigger(bot, trigger):
    ...
```

I'm not convinced that someone will want to say "this pattern will match if and only if it's plain while the second pattern will match only if and only if raw" on the same plugin callable... while having pattern that could match on the wrong version.

So yeah, that one last solution is my best offer.

**Comment by dgw** [comment]
Postponed for rework of internals to make this and other rule-related features work better. You can see some of the discussion [here](https://github.com/sopel-irc/sopel/pull/2288#discussion_r920529011) (the rest was on IRC).
